### DISCUSSION TITLE
How to customize annotation?

### DISCUSSION THREAD

**Discussion by janryWang** [mobxjs/mobx#2808] [Q&A]
I now require a deep observable object to exclude some special data types of objects, but Mobx does not provide the corresponding API, so I hope to be able to customize the annotation.

Like this.

```ts
import { makeObservable } from 'mobx'

const myAnnotation = ()=>{
  //...
}

makeObservable({
  data:{
    node:<div>hello world</div> //If the node attribute is react node(React.isValidElement),
                                // then use observable.ref, otherwise use deep observable
  }
},{
  data:myAnnotation
})
```

**Comment by iChenLei** [comment]
@urugator Any suggestion? 👀

**Reply by urugator** [reply]
Don't rely on auto conversion, expose an action `setData` (or setter) and create the observable manually there.

**Reply by janryWang** [reply]
This issue is the same issue as https://github.com/mobxjs/mobx/discussions/2809 @urugator My core requirement is to observe the properties of deep objects through customized means.

**Reply by urugator** [reply]
You're supposed to modify observables via actions - meaning you're in control of the conversion process if the automatic one doesn't work you. So as I said you can use lodash or something to traverse provided object and make things observable as required.

**Reply by janryWang** [reply]
How to convert it? Can you give a demo?

**Reply by urugator** [reply]
Another way to handle this is to wrap these in extra observable box with deep false
`store.data.node = observable.box(reactElement, { deep: false })`
Since the box is already observable, the mobx will ignore it and leaves `reactElement` untouched.

**Reply by janryWang** [reply]
The problem is that in my scenario, I don’t know whether the data type of node is a react element or a normal object. If it is a normal object, I want to continue to observe recursively. like this https://github.com/mobxjs/mobx/discussions/2809#discussioncomment-376642
